scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,306 @@
+                               (#(for-map [change %] (:cluster-name change) (:config change)))
+                               (merge current-configs))
+        ;TODO check that number of running clusters is not less than a configured value. and make no changes if so
+        ]
+    (-> changes
+      (check-for-unique-constraint-violations resulting-configs :base-path)
+      (check-for-unique-constraint-violations resulting-configs :ca-cert))))
+
+;TODO see if this is ok or need a better way
+(def scheduler-promise (promise))
+(def exit-code-syncer-state-promise (promise))
+
+(defn add-new-cluster!
+  "Add a new cluster from a dynamic cluster config"
+  [conn config]
+  (try
+    (let [config-from-template ((config/compute-cluster-templates) (:template config))
+          factory-fn (:factory-fn config-from-template)
+          resolved (cook.util/lazy-load-var factory-fn)
+          config (merge (:config config-from-template) config)
+          cluster (resolved config {:exit-code-syncer-state @exit-code-syncer-state-promise})]
+      (initialize-cluster cluster (:pool-name->fenzo @scheduler-promise)))
+    (let [temp-db-id (d/tempid :db.part/user)]
+      @(d/transact
+         conn
+         [(assoc (compute-cluster-config->compute-cluster-config-ent config) :db/id temp-db-id)]))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+(defn update-cluster!
+  "Update a cluster with a dynamic cluster config"
+  [conn {:keys [name state state-locked?] :as config} current-db-config-ents current-in-mem-configs]
+  (try
+    (let [{:keys [state-atom state-locked?-atom] :as cluster} (@cluster-name->compute-cluster-atom name)
+          current-db-config-ent (current-db-config-ents name)
+          current-in-mem-config (current-in-mem-configs name)]
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
+        (throw (ex-info "We know an in-memory config but we don't know the corresponding in-memory cluster.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (when (and state-atom state-locked?-atom)
+        (reset! state-atom state)
+        (reset! state-locked?-atom state-locked?))
+      (when current-db-config-ent
+        @(d/transact
+           conn
+           [(assoc (compute-cluster-config->compute-cluster-config-ent config) :db/id (:db/id current-db-config-ent))])))
+    {:update-succeeded true}
+    (catch Throwable t
+      (log/error t "Failed to update cluster" config)
+      {:update-succeeded false :error-message (.toString t)})))
+
+; TODO make sure anything in "locking" path times out. like db calls
+; TODO do alerts in callers of this function on return value
+(defn update-compute-clusters
```

Comment:
Maybe call it 'replace'?

Reply:
well, this can just do an add. open to naming this particular function differently